dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -57,81 +58,107 @@ void FineGrainedShuffleWriter<StreamWriterPtr>::finishWrite()
     }
 }
 
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::prepare(const Block & sample_block)
+{
+    header = sample_block.cloneEmpty();
+    num_columns = header.columns();
+    // fine_grained_shuffle_stream_count is in (0, 1024], and partition_num is uint16_t, so will not overflow.
+    num_bucket = partition_num * fine_grained_shuffle_stream_count;
+    ;
+    partition_key_containers_for_reuse.resize(collators.size());
+    resetScatterColumns();
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::flush()
+{
+    if (rows_in_blocks > 0)
+        batchWriteFineGrainedShuffle<false>();
+}
+
 template <class StreamWriterPtr>
 void FineGrainedShuffleWriter<StreamWriterPtr>::write(const Block & block)
 {
     RUNTIME_CHECK_MSG(
         block.columns() == dag_context.result_field_types.size(),
         "Output column size mismatch with field type size");
+
     size_t rows = block.rows();
     rows_in_blocks += rows;
     if (rows > 0)
     {
         blocks.push_back(block);
     }
 
-    if (static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size)
+    if (blocks.size() == fine_grained_shuffle_stream_count || static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size * fine_grained_shuffle_stream_count)
         batchWriteFineGrainedShuffle<false>();
 }
 
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::resetScatterColumns()
+{
+    scattered.resize(num_columns);
+    for (size_t col_id = 0; col_id < num_columns; ++col_id)
+    {
+        auto & column = header.getByPosition(col_id).column;
+
+        scattered[col_id].reserve(num_bucket);
+        for (size_t chunk_id = 0; chunk_id < num_bucket; ++chunk_id)
+        {
+            scattered[col_id].emplace_back(column->cloneEmpty());
+            scattered[col_id][chunk_id]->reserve(1024);
+        }
+    }
+}
+
 template <class StreamWriterPtr>
 template <bool send_exec_summary_at_last>
 void FineGrainedShuffleWriter<StreamWriterPtr>::batchWriteFineGrainedShuffle()
 {
     std::vector<TrackedMppDataPacket> tracked_packets(partition_num);
-
     if (!blocks.empty())
     {
         assert(rows_in_blocks > 0);
         assert(fine_grained_shuffle_stream_count <= 1024);
 
-        // fine_grained_shuffle_stream_count is in (0, 1024], and partition_num is uint16_t, so will not overflow.
-        uint32_t bucket_num = partition_num * fine_grained_shuffle_stream_count;
-
         HashBaseWriterHelper::materializeBlocks(blocks);
```

Comment:
from https://github.com/pingcap/tiflash/pull/6173/files#diff-ad533224c30a2d203c877d1c8eaa03cc0bcacef358a8169cf5f8915fe7970265R100-R113,
it seems that we don't need to call `HashBaseWriterHelper::materializeBlocks`.

Reply:
Materialize constant columns, seems work for ser/deser columns, not for scatter/insertRangeFrom:
https://github.com/pingcap/tiflash/blob/40ad20c9439b18221297ed2b87d9d7421d17b174/dbms/src/Flash/Coprocessor/CHBlockChunkCodec.cpp#L70

Reply:
yes, you are right.